File: subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> path;
        dfs(nums, path, 0);
        return res;
    }
    
    void dfs(vector<int>& nums, vector<int> &path, int i) {
        if (i == nums.size()) {
            res.push_back(path);
            return;
        }
        path.push_back(nums[i]);
        dfs(nums, path, i+1);
        path.pop_back();
        dfs(nums, path, i+1);
    }
private:
    vector<vector<int>> res;
};

File: python/subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        self.res = []
        self.dfs(nums,0, [])
        return self.res
        
    def dfs(self, nums, i, tmp):
        if i == len(nums):
            self.res.append(tmp[:])
            return
        tmp.append(nums[i])
        self.dfs(nums, i+1, tmp)
        tmp.pop()
        self.dfs(nums, i+1, tmp)
        
